File: electron/live-engine.ts
import { createHash } from 'node:crypto';
import type { LiveChannel } from './types.js';

function stableId(sourceId: string, group: string, name: string): string {
  return `live-${createHash('sha1').update(`${sourceId}\n${group}\n${name}`).digest('hex').slice(0, 16)}`;
}

function attributes(line: string): Record<string, string> {
  const result: Record<string, string> = {};
  for (const match of line.matchAll(/([\w-]+)="([^"]*)"/g)) result[match[1].toLowerCase()] = match[2];
  return result;
}

function validStreamUrls(value: string): string[] {
  return value
    .split(/#(?=https?:\/\/)/i)
    .map((entry) => entry.trim().replace(/^#/, ''))
    .filter((entry) => /^https?:\/\//i.test(entry));
}

function addChannel(channels: LiveChannel[], channel: LiveChannel): void {
  const existing = channels.find((entry) => entry.group === channel.group && entry.name === channel.name);
  if (!existing) {
    channels.push(channel);
    return;
  }
  const urls = [...new Set([...(existing.urls ?? [existing.url]), ...(channel.urls ?? [channel.url])])];
  existing.urls = urls;
  existing.url = urls[0];
  if (!existing.logo && channel.logo) existing.logo = channel.logo;
  if (!existing.tvgId && channel.tvgId) existing.tvgId = channel.tvgId;
}

export function parseLivePlaylist(content: string, sourceId: string, sourceName: string): LiveChannel[] {
  const lines = content.replace(/^\uFEFF/, '').split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
  const channels: LiveChannel[] = [];
  let group = sourceName || '未分组';

  for (let index = 0; index < lines.length; index++) {
    const line = lines[index];
    if (line.startsWith('#EXTINF:')) {
      const attrs = attributes(line);
      const commaAt = line.lastIndexOf(',');
      const name = (commaAt >= 0 ? line.slice(commaAt + 1) : attrs['tvg-name']).trim();
      const next = lines[index + 1];
      const urls = next && !next.startsWith('#') ? validStreamUrls(next) : [];
      if (name && urls.length) {
        const channelGroup = attrs['group-title'] || group;
        addChannel(channels, {
          id: stableId(sourceId, channelGroup, name), sourceId, sourceName,
          tvgId: attrs['tvg-id'], name, group: channelGroup, logo: attrs['tvg-logo'],
          url: urls[0], urls,
        });
        index++;
      }
      continue;
    }
    if (line.startsWith('#')) continue;
    const commaAt = line.indexOf(',');
    if (commaAt < 0) continue;
    const name = line.slice(0, commaAt).trim();
    const value = line.slice(commaAt + 1).trim();
    if (/^#genre#$/i.test(value)) {
      group = name || group;
      continue;
    }
    const urls = validStreamUrls(value);
    if (!name || !urls.length) continue;
    addChannel(channels, {
      id: stableId(sourceId, group, name), sourceId, sourceName,
      name, group, url: urls[0], urls,
    });
  }
  return channels;
}

export function mergeLiveChannels(existing: LiveChannel[], incoming: LiveChannel[]): LiveChannel[] {
  const merged = existing.map((channel) => structuredClone(channel));
  incoming.forEach((channel) => addChannel(merged, structuredClone(channel)));
  return merged;
}

File: mobile/lib/services/live_engine.dart
import 'dart:convert';

import '../models/models.dart';

class LiveEngine {
  List<LiveChannel> parseLivePlaylist(
    String content,
    String sourceId,
    String sourceName,
  ) {
    final normalized = content.isNotEmpty && content.codeUnitAt(0) == 0xFEFF
        ? content.substring(1)
        : content;
    final lines = normalized
        .split(RegExp(r'\r?\n'))
        .map((line) => line.trim())
        .where((line) => line.isNotEmpty)
        .toList();
    final channels = <LiveChannel>[];
    var group = sourceName.isEmpty ? '未分组' : sourceName;

    for (var index = 0; index < lines.length; index++) {
      final line = lines[index];
      if (line.startsWith('#EXTINF:')) {
        final attrs = _attributes(line);
        final commaAt = line.lastIndexOf(',');
        final name = (commaAt >= 0
                ? line.substring(commaAt + 1)
                : attrs['tvg-name'] ?? '')
            .trim();
        final next = index + 1 < lines.length ? lines[index + 1] : '';
        final urls = next.isNotEmpty && !next.startsWith('#')
            ? _validStreamUrls(next)
            : <String>[];
        if (name.isNotEmpty && urls.isNotEmpty) {
          final channelGroup = attrs['group-title'] ?? group;
          _addChannel(
            channels,
            LiveChannel(
              id: _stableId(sourceId, channelGroup, name),
              sourceId: sourceId,
              sourceName: sourceName,
              name: name,
              group: channelGroup,
              logo: attrs['tvg-logo'],
              url: urls.first,
              urls: urls,
            ),
          );
          index++;
        }
        continue;
      }
      if (line.startsWith('#')) continue;
      final commaAt = line.indexOf(',');
      if (commaAt < 0) continue;
      final name = line.substring(0, commaAt).trim();
      final value = line.substring(commaAt + 1).trim();
      if (RegExp(r'^#genre#$', caseSensitive: false).hasMatch(value)) {
        if (name.isNotEmpty) group = name;
        continue;
      }
      final urls = _validStreamUrls(value);
      if (name.isEmpty || urls.isEmpty) continue;
      _addChannel(
        channels,
        LiveChannel(
          id: _stableId(sourceId, group, name),
          sourceId: sourceId,
          sourceName: sourceName,
          name: name,
          group: group,
          url: urls.first,
          urls: urls,
        ),
      );
    }
    return channels;
  }

  Map<String, String> _attributes(String line) {
    final result = <String, String>{};
    for (final match in RegExp(r'([\w-]+)="([^"]*)"').allMatches(line)) {
      result[match.group(1)!.toLowerCase()] = match.group(2)!;
    }
    return result;
  }

  List<String> _validStreamUrls(String value) => value
      .split(RegExp(r'#(?=https?://)', caseSensitive: false))
      .map((entry) => entry.trim().replaceFirst(RegExp(r'^#'), ''))
      .where((entry) => RegExp(r'^https?://', caseSensitive: false).hasMatch(entry))
      .toList();

  void _addChannel(List<LiveChannel> channels, LiveChannel channel) {
    final index = channels.indexWhere(
      (entry) => entry.group == channel.group && entry.name == channel.name,
    );
    if (index < 0) {
      channels.add(channel);
      return;
    }
    final existing = channels[index];
    final urls = <String>{
      ...(existing.urls ?? [existing.url]),
      ...(channel.urls ?? [channel.url]),
    }.toList();
    channels[index] = LiveChannel(
      id: existing.id,
      sourceId: existing.sourceId,
      sourceName: existing.sourceName,
      name: existing.name,
      group: existing.group,
      url: urls.first,
      urls: urls,
      logo: existing.logo?.isNotEmpty == true ? existing.logo : channel.logo,
    );
  }

  String _stableId(String sourceId, String group, String name) {
    final encoded = base64Url.encode(utf8.encode('$sourceId\n$group\n$name'));
    return 'live-${encoded.replaceAll('=', '').substring(0, encoded.length.clamp(0, 22))}';
  }
}
